### ISSUE TITLE
Improve data transfer from server to client so it's possible for specific scripts to declare the data they need.

### ISSUE THREAD

**Issue by nerrad** [woocommerce/woocommerce-gutenberg-products-block#904]
Now that #903 has landed.  I want to take the next step of implementing a mechanism for scripts to easily declare data they need to pass through to the server.  An example where this will be useful is frontend implementation of the [All Products block](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/898) where the initial view would be populated by a json object passed along on initial render.

We don't want all data to be in ever block render, only blocks that need specific data.  So work resolving this issue will involve:

- slightly refactoring Asset management so everything isn't contained within `Assets.php`. It's handling a lot of different concerns right now and I'd like to split it out a bit (also look into automatically detecting and registering block scripts).

- introduce an api for registering what data should be loaded with a script that will automatically get enqueued with the `wc-block-settings` script and loaded to the `wc.blockSettings` global. The api might look something like this:

```php
// BlockAsset object possibly
$registered_script->add_data( 'myData', [ 'foo' => 'bar' ] );
```

And then accessible in js via

```js
// { foo: bar };
console.log( wc.blockSettings.myData );
```

Once this api is in place, we can take a look at the existing data we're passing through and see if it is truly "global" data (i.e. any block might need it) or it can be restricted to just specific blocks.  

This is just a very rough concept at the moment and I'm mostly documenting the basics so I have something to go by when I pick this up.

**Comment by nerrad** [comment]
This is available now via `AssetDataRegistry` introduced in #956
